llvm/include/llvm/IR/RepoTicket.h

```diff
@@ -167,111 +167,151 @@ class TicketNode : public MDNode {
   friend class LLVMContextImpl;
   friend class MDNode;
 
-  struct CheckLinkageType {
-    using LT = std::underlying_type<GlobalValue::LinkageTypes>::type;
-    using ULT = std::make_unsigned<LT>::type;
+  template <unsigned Bitwidth, typename Type> struct CheckType {
+    using LT = typename std::underlying_type<Type>::type;
+    using ULT = typename std::make_unsigned<LT>::type;
     template <typename U> static constexpr U max() {
       return std::numeric_limits<U>::max();
     }
-    static constexpr bool isSafeCast() {
+    static bool isSafeCast(Type Value) {
       return (std::is_unsigned<LT>::value ? true : max<LT>() > 0) &&
-             max<ULT>() <= max<unsigned>();
+             max<ULT>() <= max<unsigned>() &&
+             static_cast<unsigned>(Value) <= (1u << Bitwidth);
+    }
+  };
+
+  struct Data32 {
+    static unsigned combine(GlobalValue::VisibilityTypes high,
+                            GlobalValue::LinkageTypes low) {
+      return static_cast<unsigned>(static_cast<uint64_t>(high) << 16 |
+                                   static_cast<uint64_t>(low));
```

Comment:
Why do we cast to `uint64_t` rather than `unsigned` here? In fact, the name of the struct is `Data32`, so should we be using `uint32_t` rather than unsigned?

Reply:
> 
> 
> Why do we cast to `uint64_t` rather than `unsigned` here? 
>
Yes, you are right. Changed to `unsigned`.
>
>In fact, the name of the struct is `Data32`, so should we be using `uint32_t` rather than unsigned?
>

Since the return value is assigned to the `SubclassData32 ( SubclassData32 = Data32::combine(Visibility, Linkage);`. The `SubclassData32` is an `unsigned` type, therefore the `unsigned` is used here.

Reply:
> SubclassData32 is an unsigned type

Well that's weird. The name strongly suggests that it’s a 32-bit value rather than whatever `unsigned` happens to be…

Reply:
> 
> 
> > SubclassData32 is an unsigned type
> 
> Well that's weird. The name strongly suggests that it’s a 32-bit value rather than whatever `unsigned` happens to be…

Yes, you are right. It is weird.
However, the `SubclassData32` is defined in the class `Metadata` (not our metadata class), we can't modify it since it is used by other Metadata classes (for example, `DebugInfoMetadata`).